service/history/shard/coordinator.go

```diff
@@ -0,0 +1,214 @@
+		ReceiveFailoverMarkers(shardID int, markers []*replicator.FailoverMarkerAttributes)
+	}
+
+	coordinatorImpl struct {
+		status       int32
+		recorder     map[string]*failoverRecord
+		sendChan     chan *request
+		receiveChan  chan *request
+		shutdownChan chan struct{}
+
+		config *config.Config
+		logger log.Logger
+	}
+
+	request struct {
+		shardID int
+		markers []*replicator.FailoverMarkerAttributes
+		respCh  chan error
+	}
+
+	failoverRecord struct {
+		failoverVersion int64
+		shards          map[int]struct{}
+	}
+)
+
+// NewCoordinator initialize a failover coordinator
+func NewCoordinator(
+	config *config.Config,
+	logger log.Logger,
+) Coordinator {
+	return &coordinatorImpl{
+		status:       common.DaemonStatusInitialized,
+		recorder:     make(map[string]*failoverRecord),
+		sendChan:     make(chan *request, sendChanBufferSize),
+		receiveChan:  make(chan *request, receiveChanBufferSize),
+		shutdownChan: make(chan struct{}),
+		config:       config,
+		logger:       logger,
+	}
+}
+
+func (c *coordinatorImpl) Start() {
+
+	if !atomic.CompareAndSwapInt32(&c.status, common.DaemonStatusInitialized, common.DaemonStatusStarted) {
+		return
+	}
+
+	go c.receiveFailoverMarkersLoop()
+	go c.sendFailoverMarkerLoop()
+}
+
+func (c *coordinatorImpl) Stop() {
+
+	if !atomic.CompareAndSwapInt32(&c.status, common.DaemonStatusStarted, common.DaemonStatusStopped) {
+		return
+	}
+
+	close(c.shutdownChan)
+}
+
+func (c *coordinatorImpl) SendFailoverMarkers(
+	shardID int,
+	markers []*replicator.FailoverMarkerAttributes,
+) <-chan error {
+
+	respCh := make(chan error, 1)
+	c.sendChan <- &request{
+		shardID: shardID,
+		markers: markers,
+		respCh:  respCh,
+	}
+
+	return respCh
+}
+
+func (c *coordinatorImpl) ReceiveFailoverMarkers(
+	shardID int,
+	markers []*replicator.FailoverMarkerAttributes,
+) {
+
+	c.receiveChan <- &request{
+		shardID: shardID,
+		markers: markers,
+	}
+}
+
+func (c *coordinatorImpl) receiveFailoverMarkersLoop() {
+
+	for {
+		select {
+		case <-c.shutdownChan:
+			return
+		case request := <-c.receiveChan:
+			for _, marker := range request.markers {
+				domainID := marker.GetDomainID()
+
+				if record, ok := c.recorder[domainID]; ok {
+					// if the local failover version is smaller than the new received marker,
+					// it means there is another failover happened and the local one should be invalid.
+					if record.failoverVersion < marker.GetFailoverVersion() {
+						delete(c.recorder, domainID)
+					}
+				}
+
+				if _, ok := c.recorder[domainID]; !ok {
+					// initialize the failover record
+					c.recorder[marker.GetDomainID()] = &failoverRecord{
+						failoverVersion: marker.GetFailoverVersion(),
+						shards:          make(map[int]struct{}),
+					}
+					return
+				}
+
+				record := c.recorder[domainID]
+				record.shards[request.shardID] = struct{}{}
+				if len(record.shards) == c.config.NumberOfShards {
+					// TODO: update domain from pend_active to active
```

Comment:
what if this `update domain` failed?

Reply:
It will keep retrying until the failover timeout